inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_non_max_suppression.cpp

```diff
@@ -4,39 +4,50 @@
 #include "vpu/ngraph/operations/dynamic_shape_resolver.hpp"
-#include "vpu/ngraph/operations/dynamic_non_max_suppression.hpp"
 #include "vpu/ngraph/utilities.hpp"
-#include <vpu/utils/error.hpp>
 
+#include <vpu/utils/error.hpp>
 #include "ngraph/graph_util.hpp"
-#include "ngraph/opsets/opset3.hpp"
+#include "ngraph/opsets/opset5.hpp"
 
 #include <memory>
-#include <vpu/ngraph/operations/static_shape_non_maximum_suppression.hpp>
 
 namespace vpu {
 
 void dynamicToStaticNonMaxSuppression(std::shared_ptr<ngraph::Node> node) {
-    auto nms_dynamic = std::dynamic_pointer_cast<ngraph::vpu::op::DynamicNonMaxSuppression>(node);
-    VPU_THROW_UNLESS(nms_dynamic, "dynamicToStaticNonMaxSuppression transformation for {} of type {} expects {} as node for replacement",
-                     node->get_friendly_name(), node->get_type_info(), ngraph::vpu::op::DynamicNonMaxSuppression::type_info);
+    auto nms = std::dynamic_pointer_cast<ngraph::opset5::NonMaxSuppression>(node);
+    VPU_THROW_UNLESS(nms, "dynamicToStaticNonMaxSuppression transformation for {} of type {} expects {} as node for replacement",
+                     node->get_friendly_name(), node->get_type_info(), ngraph::opset5::NonMaxSuppression::type_info);
 
     auto staticShapeNMS = std::make_shared<ngraph::vpu::op::StaticShapeNonMaxSuppression>(
-            nms_dynamic->input_value(0),
-            nms_dynamic->input_value(1),
-            nms_dynamic->input_value(2),
-            nms_dynamic->input_value(3),
-            nms_dynamic->input_value(4),
-            nms_dynamic->get_box_encoding(),
-            nms_dynamic->get_sort_result_descending(),
-            nms_dynamic->get_output_type());
-
-    auto dynamicShapeResolver = std::make_shared<ngraph::vpu::op::DynamicShapeResolver>(
-            staticShapeNMS->output(0), staticShapeNMS->output(1));
-    dynamicShapeResolver->set_friendly_name(nms_dynamic->get_friendly_name());
-
-    ngraph::replace_node(std::move(nms_dynamic), std::move(dynamicShapeResolver));
+            nms->input_value(0),
+            nms->input_value(1),
+            nms->input_value(2),
+            nms->input_value(3),
+            nms->input_value(4),
+            nms->input_value(5),
+            nms->get_box_encoding() == ngraph::opset5::NonMaxSuppression::BoxEncodingType::CENTER ? 1 : 0,
+            nms->get_sort_result_descending(),
+            nms->get_output_type());
+
+    auto dsrIndices = std::make_shared<ngraph::vpu::op::DynamicShapeResolver>(
+            staticShapeNMS->output(0), staticShapeNMS->output(2));
+    auto dsrScores = std::make_shared<ngraph::vpu::op::DynamicShapeResolver>(
+            staticShapeNMS->output(1), staticShapeNMS->output(2));
+    dsrIndices->set_friendly_name(nms->output(0).get_node_shared_ptr()->get_friendly_name());
+    dsrScores->set_friendly_name(nms->output(1).get_node_shared_ptr()->get_friendly_name());
+
+    const auto gatherValidOutputs = std::make_shared<ngraph::opset5::Gather>(
```

Comment:
Do you introduce gather even in case if it's not needed?

Reply:
In case if the `valid_outputs` is not used in the network, there will be no `Gather` in the final model. I don't know at which moment this Gather is eliminated, but I have checked and we not receive it in the plugin.

Reply:
Let's not rely on logic we don't even understand completely. If you know there are cases when `Gather` is not needed - let's not introduce it

Reply:
I believe this is a way we actually should work with NGraph, and we already had an email thread on this (but unfortunately I can't find it).
@GlebKazantaev could you please help us there? Here I need to add an additional node for some of outputs. Do I need to check if this output is used and add an additional node only in this case, or it's fine to always add this additional layer just because it will be eliminated later if it's not used. What is the best practice?

Reply:
Well, I've investigated this myself and I completely sure that this checks are not necessary. In case no one used the third output of NonMaxSuppression, this Gather will be destroyed right after this function call ends. In NGraph nodes are holding their output nodes pointers as non-owning pointers, while the input nodes pointers are owning. In case `nms->output(2)` has no children, there will be no actual replacement, and none of nodes will hold this Gather as its input (while StaticShapeNMS hold it as its output but in non-owning way), that's why the Gather will be destroyed and it's by NGraph design.

Reply:
Ok, thanks for investigation

But still, I'm not sure it's a good idea to create entity just to be destroyed. Not performant specifically, you know..

Reply:
BTW, do I understand correctly, that if NMS 2nd output is used in the network, so adding Gather you just adding extra output consumer, so Gather's never gonna be consumed by some operation here?

Reply:
I thought about it, but otherwise the code will be covered by a huge number of `if` branches, for example in this pass we may  check this for all 3 outputs to avoid the creation of both DSRs and Gather, but this feature allow us not to do so and let the code stay clean and readable.

Reply:
If you mean the 3rd output of NMS (which is number of valid boxes), so its consumers will then consume the Gather which will gather the first element of StaticShapeNMS output shape (because the StaticShapeNMS gives us not just valid boxes number but the shape in form of [valid_boxes_number, 3]).

Reply:
> I thought about it, but otherwise the code will be covered by a huge number of if branches, for example in this pass we may check this for all 3 outputs to avoid the creation of both DSRs and Gather, but this feature allow us not to do so and let the code stay clean and readable.

The fact is to understand how it works you had to spend some time on deep investigation, only after that you're sure there is nothing bad with extra ngraph nodes. It's clearly not clean and readable code :)

Anyway, since this is the only issue I'm okay to leave it as is